airbyte-integrations/bases/base-python/base_python/sdk/abstract_source.py

```diff
@@ -0,0 +1,139 @@
+        """Discover streams"""
+        streams = [stream.as_airbyte_stream() for stream in self.streams(config=config)]
+        return AirbyteCatalog(streams=streams)
+
+    def check(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> AirbyteConnectionStatus:
+        """Check connection"""
+        check_succeeded, error = self.check_connection(logger, config)
+        if not check_succeeded:
+            return AirbyteConnectionStatus(status=Status.FAILED, message=str(error))
+
+        return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+
+    def read(
+            self, logger: AirbyteLogger, config: Mapping[str, Any], catalog: ConfiguredAirbyteCatalog, state: MutableMapping[str, Any] = None
+    ) -> Iterator[AirbyteMessage]:
+
+        state = state or {}
+        total_state = copy.deepcopy(state)
+        logger.info(f"Starting syncing {self.name}")
+        # TODO assert all streams exist in the connector
+        # get the streams once in case the connector needs to make any queries to generate them
+        stream_instances = {s.name: s for s in self.streams(config)}
+        for configured_stream in catalog.streams:
+            try:
+                stream_instance = stream_instances[configured_stream.stream.name]
+                yield from self._read_stream(logger=logger, stream_instance=stream_instance, configured_stream=configured_stream,
+                                             state=total_state)
+            except Exception as e:
+                logger.exception(f"Encountered an exception while reading stream {self.name}")
+                raise e
+
+        logger.info(f"Finished syncing {self.name}")
+
+    def _read_stream(
+            self,
+            logger: AirbyteLogger,
+            stream_instance: Stream,
+            configured_stream: ConfiguredAirbyteStream,
+            state: MutableMapping[str, Any]
+    ) -> Iterator[AirbyteMessage]:
+        stream_name = configured_stream.stream.name
+        use_incremental = configured_stream.sync_mode == SyncMode.incremental and stream_instance.supports_incremental
+
+        stream_state = {}
+        if use_incremental and state.get(stream_name):
+            logger.info(f"Set state of {stream_name} stream to {state.get(stream_name)}")
+            stream_state = state.get(stream_name)
+
+        logger.info(f"Syncing stream: {stream_name} ")
+        record_counter = 0
+        for record in stream_instance.read_stream(stream_state=copy.deepcopy(stream_state)):
+            now_millis = int(datetime.now().timestamp()) * 1000
+            message = AirbyteRecordMessage(stream=stream_name, data=record, emitted_at=now_millis)
+            yield AirbyteMessage(type=MessageType.RECORD, record=message)
+
+            record_counter += 1
+            if use_incremental:
+                stream_state = stream_instance.get_updated_state(stream_state, record)
+                if record_counter % stream_instance.state_checkpoint_interval == 0:
```

Comment:
I wonder if we should hide the logic so it it is possible to add more stream logic in the future

`should_checkpoint(counter, current_record)`

Reply:
maybe? I think at this point I would love to merge this and start testing. 

This seems like a backwards-compatible change to make in the future so maybe let's wait till we have an actual use case in front of us to start doing it?

Reply:
Yup!